services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/NeutronRouterServiceImpl.java

```diff
@@ -0,0 +1,484 @@
+            for (RouteEntry routeEntry : routeEntities) {
+                String nextHop = routeEntry.getNexthop();
+                if (gatewayIp.equals(nextHop)) {
+                    throw new RouterInterfaceAreUsedByRoutes();
+                }
+            }
+
+        }
+
+        // remove interface
+        subnet.setAttachedRouterId("");
+
+        List<String> ports = router.getPorts();
+        if (ports == null){
+            return new RouterInterfaceResponse();
+        }
+
+        if (ports.contains(portId)) {
+            ports.remove(portId);
+        }
+        router.setPorts(ports);
+        this.routerDatabaseService.addRouter(router);
+
+        // update device_id and device_owner
+        PortEntity portEntity = new PortEntity();
+        portEntity.setDeviceId(null);
+        portEntity.setDeviceOwner(null);
+        subnet.setPort(portEntity);
+
+        // update subnet
+        this.routerToSubnetService.updateSubnet(projectId, subnetid, subnet);
+
+        // Construct response
+        List<String> subnetIds = new ArrayList<>(){};
+        subnetIds.add(subnetid);
+        return new RouterInterfaceResponse(routerId, subnet.getVpcId(), portId, subnetid, subnetIds, projectId, projectId, subnet.getTags());
+
+    }
+
+    @Override
+    public RoutesToNeutronWebResponse addRoutesToNeutronRouter(String routerid, RoutesToNeutronRouterRequestObject requestRouter) throws ResourceNotFoundException, ResourcePersistenceException, RouterOrSubnetAndPortNotExistOrNotVisible, DatabasePersistenceException, DestinationOrNexthopCanNotBeNull {
+        RoutesToNeutronRouterResponseObject responseRouter = new RoutesToNeutronRouterResponseObject();
+        List<RoutesToNeutronRouteObject> responseRoutes = new ArrayList<>();
+
+        Router router = this.routerDatabaseService.getByRouterId(routerid);
+        if (router == null) {
+            throw new RouterOrSubnetAndPortNotExistOrNotVisible();
+        }
+        RouteTable routeTable = router.getRouteTable();
+        List<RouteEntry> routeEntities = routeTable.getRouteEntities();
+
+
+        List<RoutesToNeutronRouteObject> requestRoutes = requestRouter.getRoutes();
+        for (RoutesToNeutronRouteObject requestRoute : requestRoutes) {
+            boolean isExit = false;
+            String requestDestination = requestRoute.getDestination();
+            String requestNexthop = requestRoute.getNexthop();
+
+            if (requestDestination == null || requestNexthop == null) {
+                throw new DestinationOrNexthopCanNotBeNull();
+            }
+
+            for (RouteEntry routeEntry : routeEntities) {
+                String destination = routeEntry.getDestination();
+                String nexthop = routeEntry.getNexthop();
+                if (destination.equals(requestDestination) && nexthop.equals(requestNexthop)) {
+                    isExit = true;
+                    break;
+                }
```

Comment:
@kevin-zhonghao Here has two for-loops for matching rules, it takes O(n^2). Do we have better algorithm reducing it to linear time O(n)? Such as using destination as index or key for RouteEntry.

Reply:
If we use destination as index or key for RouteEntry, we may not guarantee the uniqueness of the RouteEntry key. Although here is two for-loops, but one of the n depends on requestbody, so the time complexity should be O(m * n), and m may not be very large, right?

Reply:
Agree. we can leave current design and implementation for now. We can wait until it hits performance issue when we doing perf-test.